x/bridge/observer/observer.go

```diff
@@ -2,118 +2,212 @@ package observer
+			tx.DstChain,
+			tx.Id,
+		))
+		return
+	}
+	cr, err := o.getConfirmationsRequired(ctx, tx)
+	if err != nil {
+		o.logger.With("error", err, "id", tx.Id).
+			Error("Failed to get confirmations required for outbound transfer")
+		return
 	}
 
+	o.outTxQueue[tx.SrcChain] = append(o.outTxQueue[tx.SrcChain], TxQueueItem{tx, cr})
+}
+
+func (o *Observer) processOutbound(ctx context.Context) {
 	for {
 		select {
 		case <-o.stopChan:
+			o.sendOutbound(ctx)
 			return
-		case event := <-txs:
-			if newBlock, ok := event.Data.(comettypes.EventDataNewBlock); ok {
-				results, err := o.cometRpc.BlockResults(ctx, &newBlock.Block.Height)
-				if err != nil {
-					o.logger.Error("Observer failed to fetch block results for block", newBlock.Block.Height)
-					continue
-				}
-
-				for _, r := range results.TxsResults {
-					if r.IsErr() {
-						continue
-					}
-					for _, e := range r.Events {
-						if _, ok := events[e.Type]; !ok {
-							continue
-						}
-						select {
-						case o.eventsOutChan <- e:
-						case <-o.stopChan:
-							o.logger.Info("Observer exiting early, event skipped: ", e)
-							return
-						}
-					}
-				}
+		case <-time.After(o.sendPeriod):
+			o.sendOutbound(ctx)
+		}
+	}
+}
+
+func (o *Observer) sendOutbound(ctx context.Context) {
+	o.outLock.Lock()
+	defer o.outLock.Unlock()
+
+	for srcId, queue := range o.outTxQueue {
+		srcChain := o.chains[srcId]
+		height, err := srcChain.Height(ctx)
+		if err != nil {
+			o.logger.Error(fmt.Sprintf("Failed to get height for %s: %s", srcId, err.Error()))
+			continue
+		}
+
+		var newQueue []TxQueueItem
+		for _, out := range queue {
+			confirmed, err := o.isTxConfirmed(ctx, height, &out)
+			if err != nil {
+				o.logger.Error(fmt.Sprintf("Failed to confirm tx %s", err.Error()))
+				newQueue = append(newQueue, out)
+				continue
+			}
+			if !confirmed {
+				newQueue = append(newQueue, out)
+				continue
+			}
+
+			err = o.chains[out.Tx.DstChain].SignalInboundTransfer(ctx, out.Tx)
+			if err != nil {
+				newQueue = append(newQueue, out)
 			}
 		}
+		o.outTxQueue[srcId] = newQueue
+	}
+}
+
+func (o *Observer) getConfirmationsRequired(ctx context.Context, tx Transfer) (uint64, error) {
+	if tx.SrcChain == ChainIdOsmosis {
+		return 1, nil
+	}
+
+	chain, ok := o.chains[ChainIdOsmosis]
+	if !ok {
+		return 0, fmt.Errorf("chain client for %s not found", ChainIdOsmosis)
+	}
+	cr, err := chain.ConfirmationsRequired(ctx, bridgetypes.AssetID{
+		SourceChain: string(tx.SrcChain),
+		Denom:       tx.Asset,
+	})
+	if err != nil {
+		return 0, err
+	}
+	return cr, nil
+}
+
+func (o *Observer) isTxConfirmed(
+	ctx context.Context,
+	curHeight uint64,
+	item *TxQueueItem,
+) (bool, error) {
+	// In Bitcoin, the number of confirmations: currentHeight - txHeight + 1, so the first block counts as 1.
+	// Therefore, count the first block as 1 for all other chains.
+	if curHeight-item.Tx.Height+1 < item.ConfirmationsRequired {
+		return false, nil
+	}
```

Comment:
I apologize if this makes things more confusing, I just think treating 0 confirmations as a valid response makes things slightly more dangerous and is worth the added complexity here. If others disagree, happy to move back (to be clear, what you did here is what I asked, I am just making it clear that I am open to changing it back if others feel it is too confusing)